lib/util/cache/repository/impl/s3.ts

```diff
@@ -0,0 +1,71 @@
+import { Readable } from 'stream';
+import {
+  GetObjectCommand,
+  GetObjectCommandInput,
+  PutObjectCommand,
+  PutObjectCommandInput,
+} from '@aws-sdk/client-s3';
+import { logger } from '../../../../logger';
+import { getS3Client, parseS3Url } from '../../../s3';
+import { streamToString } from '../../../streams';
+import type { RepoCacheRecord } from '../types';
+import { RepoCacheBase } from './base';
+
+export class RepoCacheS3 extends RepoCacheBase {
+  private readonly s3Client;
+  private readonly bucket;
+
+  constructor(repository: string, url: string) {
+    super(repository);
+    this.bucket = parseS3Url(url)?.Bucket;
+    this.s3Client = getS3Client();
+  }
+
+  async read(): Promise<string | null> {
+    const cacheFileName = this.getCacheFileName();
+    const s3Params: GetObjectCommandInput = {
+      Bucket: this.bucket,
+      Key: cacheFileName,
+    };
+    try {
+      const { Body: res } = await this.s3Client.send(
+        new GetObjectCommand(s3Params)
+      );
+      if (res instanceof Readable) {
```

Comment:
it seems `res` can be a `Blob` ?

https://github.com/renovatebot/renovate/blob/229643f7e426192ba1bcb53d5ad6cf11bf304341/lib/modules/datasource/maven/util.ts#L118-L121

Reply:
`Body: Readable | ReadableStream | Blob`

https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-s3/interfaces/getobjectcommandoutput.html#body

Reply:
im not sure node's Blob will handle DOM's blob as returned from s3.
is there some kind of conversion behind the scenes?

```
lib/util/cache/repository/impl/s3.spec.ts:43:48 - error TS2322: Type 'Blob' is not assignable to type 'Blob | ReadableStream<any> | Readable | undefined'.
  Type 'import("buffer").Blob' is not assignable to type 'Blob'.
    The types returned by 'stream()' are incompatible between these types.
      Type 'unknown' is not assignable to type 'ReadableStream<Uint8Array>'.

43     s3Mock.on(GetObjectCommand).resolvesOnce({ Body: new Blob(["{}"]) });
                                                  ~~~~

```

Reply:
you need to use `'import("buffer").Blob`

Reply:
That's exactly what I'm using, note the `Type 'import("buffer").Blob' is not assignable to type 'Blob'.` from the TS error second line.

the s3 res Body type is this [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob), while node's is entirely different. thats what causes this error.

Reply:
ok, this type only occures when using the sdk in browser, so then you'll get a `window.Blob`. But in node you only get the `Readable | ReadableStream`.

There is currently no elegant way for typescript to support suche a scenario, so i think we can simply skip the `Blob` type

Reply:
![image](https://user-images.githubusercontent.com/1798109/186348649-46fd8540-3713-4879-a669-f44213aca254.png)

😕

Reply:
![image](https://user-images.githubusercontent.com/1798109/186349657-d8d3dedc-9ad2-48ad-91f2-9c099cca3c01.png)

`steam()` function has wrong return type 😞

Reply:
https://github.com/DefinitelyTyped/DefinitelyTyped/issues/59118

Reply:
probably best solution:
```diff
diff --git a/lib/util/cache/repository/impl/s3.spec.ts b/lib/util/cache/repository/impl/s3.spec.ts
index 96c854bc0..92ee6c2af 100644
--- a/lib/util/cache/repository/impl/s3.spec.ts
+++ b/lib/util/cache/repository/impl/s3.spec.ts
@@ -1,3 +1,4 @@
+import { Blob as BufferBlob } from 'buffer';
 import { Readable } from 'stream';
 import {
   GetObjectCommand,
@@ -38,6 +39,16 @@ describe('util/cache/repository/impl/s3', () => {
     expect(logger.debug).toHaveBeenCalledWith('RepoCacheS3.read() - success');
   });

+  it('successfully reads from s3 (blob)', async () => {
+    s3Mock
+      .on(GetObjectCommand)
+      .resolvesOnce({ Body: new BufferBlob(['{}']) as Blob });
+    const s3Cache = new RepoCacheS3(repository, url);
+    await expect(s3Cache.read()).toResolve();
+    expect(logger.warn).toHaveBeenCalledTimes(0);
+    expect(logger.debug).toHaveBeenCalledWith('RepoCacheS3.read() - success');
+  });
+
   it('gets an unexpected response from s3', async () => {
     s3Mock.on(GetObjectCommand).resolvesOnce({});
     const s3Cache = new RepoCacheS3(repository, url);
```

Reply:
lets ignore blob and log a warning if it's not a `Readable`

Reply:
sure, but seperate PR